rule/context-as-argument.go

```diff
@@ -33,20 +57,48 @@ func (r *ContextAsArgumentRule) Name() string {
 }
 
 type lintContextArguments struct {
-	file      *lint.File
-	fileAst   *ast.File
-	onFailure func(lint.Failure)
+	file             *lint.File
+	fileAst          *ast.File
+	allowTypesBefore []string
+	onFailure        func(lint.Failure)
 }
 
 func (w lintContextArguments) Visit(n ast.Node) ast.Visitor {
 	fn, ok := n.(*ast.FuncDecl)
 	if !ok || len(fn.Type.Params.List) <= 1 {
 		return w
 	}
+	allowTypesLookup := make(map[string]struct{}, len(w.allowTypesBefore))
+	for _, v := range w.allowTypesBefore {
+		allowTypesLookup[v] = struct{}{}
+	}
+
+	fnArgs := fn.Type.Params.List
+	// trim off all types we've been configured to skip
+	for len(fnArgs) > 0 {
+		typeStr, ok := astExprTypeStr(fnArgs[0].Type)
+		if !ok {
+			// assume we're done. This can happen, for example, with a function type
+			// argument (`func(x func()...)` which we choose not to represent.
+			break
+		}
+		_, isAllowed := allowTypesLookup[typeStr]
+		if isAllowed {
+			// trim
+			fnArgs = fnArgs[1:]
+		} else {
+			// first non-trimmed argument means we've trimmed all prefix args
+			break
+		}
+	}
+
+	if len(fnArgs) <= 1 {
+		return w
+	}
 	// A context.Context should be the first parameter of a function.
 	// Flag any that show up after the first.
-	previousArgIsCtx := isPkgDot(fn.Type.Params.List[0].Type, "context", "Context")
-	for _, arg := range fn.Type.Params.List[1:] {
+	previousArgIsCtx := isPkgDot(fnArgs[0].Type, "context", "Context")
```

Comment:
We might simplify the check with something like:
```go
ctxIsAllowed := true
for _, arg := range fnArgs {
		argIsCtx := isPkgDot(arg.Type, "context", "Context")
		if argIsCtx && !ctxIsAllowed {
			w.onFailure( ... )
		}
		// check if the type is in the allowTypesLookup  map or is context.Context
		// if is not the case, make ctxIsAllowed = false
}
```

This lets us avoid doing all the parameter list trimming stuff